File: src/lib/api/sync.ts
import { apiFetch } from './client'

export type SyncImageItem = {
	id: string
	type: 'url' | 'file'
	url?: string
	content_base64?: string
	filename?: string
	hash?: string
}

export type PublishBlogRequest = {
	slug: string
	title: string
	md: string
	tags: string[]
	date?: string
	summary?: string
	hidden?: boolean
	category?: string
	cover?: SyncImageItem | null
	images?: SyncImageItem[]
	mode?: 'create' | 'edit'
	originalSlug?: string | null
}

export async function publishBlogToGithub(req: PublishBlogRequest): Promise<{ commit_sha: string }> {
	return apiFetch('/api/sync/publish-blog', {
		method: 'POST',
		body: JSON.stringify(req)
	})
}

export async function deleteBlogFromGithub(slug: string): Promise<{ commit_sha: string }> {
	return apiFetch('/api/sync/delete-blog', {
		method: 'POST',
		body: JSON.stringify({ slug })
	})
}

export type UploadFileItem = {
	content_base64: string
	filename: string
}

export type SaveConfigRequest = {
	siteContent: any
	cardStyles: any
	favicon?: UploadFileItem | null
	avatar?: UploadFileItem | null
	artImageUploads?: Record<string, UploadFileItem>
	removedArtImages?: Array<{ id: string; url: string }>
	backgroundImageUploads?: Record<string, UploadFileItem>
	removedBackgroundImages?: Array<{ id: string; url: string }>
	socialButtonImageUploads?: Record<string, UploadFileItem>
}

export async function saveConfigToGithub(req: SaveConfigRequest): Promise<{ commit_sha: string }> {
	return apiFetch('/api/sync/save-config', {
		method: 'POST',
		body: JSON.stringify(req)
	})
}

export type BlogIndexItem = {
	slug: string
	title: string
	type?: string
	tags: string[]
	date: string
	hidden?: boolean
	summary?: string
	cover?: string
	category?: string
}

export type BatchEditBlogsRequest = {
	removedSlugs: string[]
	nextItems: BlogIndexItem[]
	categories: string[]
}

export async function batchEditBlogsOnGithub(req: BatchEditBlogsRequest): Promise<{ commit_sha: string }> {
	return apiFetch('/api/sync/batch-edit-blogs', {
		method: 'POST',
		body: JSON.stringify(req)
	})
}

export type SaveJsonFileRequest = {
	path: string
	content: any
	commitMessage: string
}

export async function saveJsonFileToGithub(req: SaveJsonFileRequest): Promise<{ commit_sha: string }> {
	return apiFetch('/api/sync/save-json-file', {
		method: 'POST',
		body: JSON.stringify(req)
	})
}

export type CommitFileItem = {
	path: string
	content_base64: string | null
	encoding?: string
}

export type CommitRequest = {
	commitMessage: string
	files: CommitFileItem[]
}

export async function commitFilesToGithub(req: CommitRequest): Promise<{ commit_sha: string }> {
	return apiFetch('/api/sync/commit', {
		method: 'POST',
		body: JSON.stringify(req)
	})
}

File: backend/app/schemas/sync.py
from typing import Optional, List, Dict, Any
from pydantic import BaseModel

class ImageItem(BaseModel):
    id: str
    type: str  # "url" | "file"
    url: Optional[str] = None
    content_base64: Optional[str] = None
    filename: Optional[str] = None
    hash: Optional[str] = None

class PublishBlogRequest(BaseModel):
    slug: str
    title: str
    md: str
    tags: List[str]
    date: Optional[str] = None
    summary: Optional[str] = None
    hidden: Optional[bool] = False
    category: Optional[str] = None
    cover: Optional[ImageItem] = None
    images: Optional[List[ImageItem]] = []
    mode: Optional[str] = "create"
    originalSlug: Optional[str] = None

class DeleteBlogRequest(BaseModel):
    slug: str

class UploadFileItem(BaseModel):
    content_base64: str
    filename: str

class ArtImageConfig(BaseModel):
    id: str
    url: str

class BackgroundImageConfig(BaseModel):
    id: str
    url: str

class SaveConfigRequest(BaseModel):
    siteContent: dict
    cardStyles: dict
    favicon: Optional[UploadFileItem] = None
    avatar: Optional[UploadFileItem] = None
    artImageUploads: Optional[Dict[str, UploadFileItem]] = None
    removedArtImages: Optional[List[ArtImageConfig]] = None
    backgroundImageUploads: Optional[Dict[str, UploadFileItem]] = None
    removedBackgroundImages: Optional[List[BackgroundImageConfig]] = None
    socialButtonImageUploads: Optional[Dict[str, UploadFileItem]] = None

class BlogIndexItem(BaseModel):
    slug: str
    title: str
    type: Optional[str] = None
    tags: List[str]
    date: str
    hidden: Optional[bool] = False
    summary: Optional[str] = None
    cover: Optional[str] = None
    category: Optional[str] = None

class BatchEditBlogsRequest(BaseModel):
    removedSlugs: List[str]
    nextItems: List[BlogIndexItem]
    categories: List[str]

class SaveJsonFileRequest(BaseModel):
    path: str
    content: Any
    commitMessage: str

class CommitFileItem(BaseModel):
    path: str
    content_base64: Optional[str] = None  # None means delete
    encoding: Optional[str] = "base64"

class CommitRequest(BaseModel):
    commitMessage: str
    files: List[CommitFileItem]
